controllers/topolvm_node.go

```diff
@@ -0,0 +1,276 @@
+	v1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	cutil "sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"strings"
+)
+
+const (
+	topolvmNodeName = "topolvm-node"
+)
+
+type topolvmNode struct{}
+
+func (n topolvmNode) getName() string {
+	return topolvmNodeName
+}
+
+func (n topolvmNode) ensureCreated(r *LVMClusterReconciler, ctx context.Context, lvmCluster *lvmv1alpha1.LVMCluster) error {
+	nodeDaemonSet := getNodeDaemonSet(lvmCluster)
+	result, err := cutil.CreateOrUpdate(ctx, r.Client, nodeDaemonSet, func() error {
+		// make sure LVMCluster CR garbage collects this daemonset and also block owner removal
+		return cutil.SetControllerReference(lvmCluster, nodeDaemonSet, r.Scheme)
+	})
+
+	switch result {
+	case cutil.OperationResultCreated:
+		r.Log.Info(topolvmNodeName, "operation", result, "name", nodeDaemonSet.Name)
+	case cutil.OperationResultUpdated:
+		r.Log.Info(topolvmNodeName, "operation", result, "name", nodeDaemonSet.Name)
+	case cutil.OperationResultNone:
+		r.Log.Info(topolvmNodeName, "operation", result, "name", nodeDaemonSet.Name)
+	default:
+		r.Log.Error(err, fmt.Sprintf("%s reconcile failure", topolvmNodeName), "name", nodeDaemonSet.Name)
+		return err
+	}
+
+	return nil
+}
+
+// ensureDeleted should wait for the resources to be cleaned up
+func (n topolvmNode) ensureDeleted(r *LVMClusterReconciler, ctx context.Context, lvmCluster *lvmv1alpha1.LVMCluster) error {
+	NodeDaemonSet := &appsv1.DaemonSet{}
+	err := r.Client.Get(ctx, types.NamespacedName{Name: TopolvmNodeDaemonsetName, Namespace: lvmCluster.Namespace}, NodeDaemonSet)
+
+	if err != nil {
+		if errors.IsNotFound(err) {
+			r.Log.Info("topolvm node deleted", "TopolvmNode", NodeDaemonSet.Name)
+			return nil
+		}
+		r.Log.Error(err, "unable to retrieve topolvm node daemonset", "TopolvmNode", NodeDaemonSet.Name)
+		return err
+	}
+	return nil
+}
+
+// updateStatus should optionally update the CR's status about the health of the managed resource
+// each unit will have updateStatus called individually so
+// avoid status fields like lastHeartbeatTime and have a
+// status that changes only when the operands change.
+func (n topolvmNode) updateStatus(r *LVMClusterReconciler, ctx context.Context, lvmCluster *lvmv1alpha1.LVMCluster) error {
+	return nil
+}
+
+func getNodeDaemonSet(lvmCluster *lvmv1alpha1.LVMCluster) *v1.DaemonSet {
+	hostPathDirectory := corev1.HostPathDirectory
+	hostPathDirectoryOrCreateType := corev1.HostPathDirectoryOrCreate
+	storageMedium := corev1.StorageMediumMemory
+
+	volumes := []corev1.Volume{
+		{Name: "registration-dir", VolumeSource: corev1.VolumeSource{HostPath: &corev1.HostPathVolumeSource{Path: fmt.Sprintf("%splugins_registry/", getAbsoluteKubeletPath(CSIKubeletRootDir)), Type: &hostPathDirectory}}},
```

Comment:
These lines are too long to read properly. Please use a format like the following:

{field: value ,
field: value,
...
}